use active sock count to control limit on pool

adilhafeez opened the pull request:
While investigating socket timeout issue, we found out that we're opening more connections to mongo server than what we've set in poolLimit. It appears that mgo has a but under high load it can open lot more connections.

This code change attempts to put max limit on active connections to mongo server.

adilhafeez committed:
use active sock count to control limit on pool

adilhafeez committed:
fix var name

adilhafeez commented:
live socket count before and after (this test is with poolLimit of 10)


before

```
2017-10-20T23:03:49.10752 {"level":"info","msg":"done scheduling go routines","time":"2017-10-20T23:03:49Z"}
2017-10-20T23:03:49.10770 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10770 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10773 Connection to localhost:27019 established.
2017-10-20T23:03:49.10796 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10797 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10797 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10797 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10805 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10811 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10816 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.10819 Connection to localhost:27019 established.
2017-10-20T23:03:49.10834 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11004 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11005 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11005 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11005 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11006 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11006 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11006 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11006 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11006 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11007 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11007 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11007 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11007 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11007 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11008 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11008 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11008 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11008 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11009 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11009 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11013 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11013 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11013 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11013 Connection to localhost:27019 established.
2017-10-20T23:03:49.11014 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11014 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11014 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11014 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11014 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11015 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11035 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11063 Connection to localhost:27019 established.
2017-10-20T23:03:49.11072 Connection to localhost:27019 established.
2017-10-20T23:03:49.11072 Connection to localhost:27019 established.
2017-10-20T23:03:49.11073 Connection to localhost:27019 established.
2017-10-20T23:03:49.11073 Connection to localhost:27019 established.
2017-10-20T23:03:49.11079 Connection to localhost:27019 established.
2017-10-20T23:03:49.11079 Connection to localhost:27019 established.
2017-10-20T23:03:49.11090 Connection to localhost:27019 established.
2017-10-20T23:03:49.11090 Connection to localhost:27019 established.
2017-10-20T23:03:49.11091 Connection to localhost:27019 established.
2017-10-20T23:03:49.11091 Connection to localhost:27019 established.
2017-10-20T23:03:49.11099 Connection to localhost:27019 established.
2017-10-20T23:03:49.11099 Connection to localhost:27019 established.
2017-10-20T23:03:49.11100 Connection to localhost:27019 established.
2017-10-20T23:03:49.11101 Connection to localhost:27019 established.
2017-10-20T23:03:49.11108 Connection to localhost:27019 established.
2017-10-20T23:03:49.11109 Connection to localhost:27019 established.
2017-10-20T23:03:49.11120 Connection to localhost:27019 established.
2017-10-20T23:03:49.11121 Connection to localhost:27019 established.
2017-10-20T23:03:49.11121 Connection to localhost:27019 established.
2017-10-20T23:03:49.11121 Connection to localhost:27019 established.
2017-10-20T23:03:49.11122 Connection to localhost:27019 established.
2017-10-20T23:03:49.11130 Connection to localhost:27019 established.
2017-10-20T23:03:49.11131 Connection to localhost:27019 established.
2017-10-20T23:03:49.11131 Connection to localhost:27019 established.
2017-10-20T23:03:49.11141 Connection to localhost:27019 established.
2017-10-20T23:03:49.11141 Connection to localhost:27019 established.
2017-10-20T23:03:49.11141 Connection to localhost:27019 established.
2017-10-20T23:03:49.11142 Connection to localhost:27019 established.
2017-10-20T23:03:49.11152 Connection to localhost:27019 established.
2017-10-20T23:03:49.11152 Connection to localhost:27019 established.
2017-10-20T23:03:49.11152 Connection to localhost:27019 established.
2017-10-20T23:03:49.11153 Connection to localhost:27019 established.
2017-10-20T23:03:49.11164 Connection to localhost:27019 established.
2017-10-20T23:03:49.11164 Connection to localhost:27019 established.
2017-10-20T23:03:49.11198 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11225 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11225 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11225 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11226 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11229 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11230 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11231 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11231 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11273 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11276 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11276 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11382 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11386 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11387 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11387 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11388 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11388 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11388 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11389 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11389 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11389 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11393 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11394 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11394 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11394 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11396 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11397 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11397 Connection to localhost:27019 established.
2017-10-20T23:03:49.11397 Connection to localhost:27019 established.
2017-10-20T23:03:49.11397 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11397 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11403 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11403 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11403 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11404 Connection to localhost:27019 established.
2017-10-20T23:03:49.11404 Connection to localhost:27019 established.
2017-10-20T23:03:49.11404 Connection to localhost:27019 established.
2017-10-20T23:03:49.11405 Connection to localhost:27019 established.
2017-10-20T23:03:49.11406 Connection to localhost:27019 established.
2017-10-20T23:03:49.11406 Connection to localhost:27019 established.
2017-10-20T23:03:49.11406 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11406 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11406 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11407 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11407 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11407 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11407 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11407 WARNING: Per-server connection limit reached.
2017-10-20T23:03:49.11408 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:03:49.11408 Connection to localhost:27019 established.
2017-10-20T23:03:49.11438 Connection to localhost:27019 established.
2017-10-20T23:03:49.11463 Connection to localhost:27019 established.
2017-10-20T23:03:49.60829 {"MasterConns":49,"SocketRefs":0,"SocketsAlive":49,"SocketsInUse":0,"level":"info","msg":"socket stats","time":"2017-10-20T23:03:49Z"}
```
after
```
2017-10-20T23:00:34.34217 {"level":"info","msg":"done scheduling go routines","time":"2017-10-20T23:00:34Z"}
2017-10-20T23:00:34.34222 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34222 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34222 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34222 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34223 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34223 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34223 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34223 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34223 Establishing new connection to localhost:27019 (timeout=2s)...
2017-10-20T23:00:34.34224 WARNING: Per-server connection limit reached.
2017-10-20T23:00:34.34224 WARNING: Per-server connection limit reached.
2017-10-20T23:00:34.34224 WARNING: Per-server connection limit reached.
2017-10-20T23:00:34.34224 WARNING: Per-server connection limit reached.
2017-10-20T23:00:34.34224 WARNING: Per-server connection limit reached.
2017-10-20T23:00:34.34225 WARNING: Per-server connection limit reached.
...
...
2017-10-20T23:00:34.34271 Connection to localhost:27019 established.
2017-10-20T23:00:34.34271 Connection to localhost:27019 established.
2017-10-20T23:00:34.34271 Connection to localhost:27019 established.
2017-10-20T23:00:34.34271 Connection to localhost:27019 established.
2017-10-20T23:00:34.34271 Connection to localhost:27019 established.
2017-10-20T23:00:34.34272 Connection to localhost:27019 established.
2017-10-20T23:00:34.34272 Connection to localhost:27019 established.
2017-10-20T23:00:34.34272 Connection to localhost:27019 established.
2017-10-20T23:00:34.34272 Connection to localhost:27019 established.
2017-10-20T23:00:34.83959 {"MasterConns":9,"SocketRefs":0,"SocketsAlive":9,"SocketsInUse":0,"level":"info","msg":"socket stats","time":"2017-10-20T23:00:34Z"}
2017-10-20T23:00:35.34018 {"MasterConns":9,"SocketRefs":0,"SocketsAlive":9,"SocketsInUse":0,"level":"info","msg":"socket stats","time":"2017-10-20T23:00:35Z"}
2017-10-20T23:00:35.84078 {"MasterConns":9,"SocketRefs":0,"SocketsAlive":9,"SocketsInUse":0,"level":"info","msg":"socket stats","time":"2017-10-20T23:00:35Z"}
```

adilhafeez committed:
fix condition

rodaine commented on server.go:
```diff
@@ -56,6 +56,7 @@ type mongoServer struct {
 	pingWindow         [6]time.Duration
 	info               *mongoServerInfo
 	maxSocketReuseTime time.Duration
+	activeSocketsCount int
```
Isn't this equivalent to `len(server.liveSockets)`? This line seems to imply so: https://github.com/lyft/mgo/pull/6/files#diff-34c6b408d72845d076d47126c29948d1R255

If so, can we not track it separately and instead compute it?

adilhafeez replied:
We're using activeSocketsCount as a placeholder to count `liveSockets` var.

Basically at [this](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L131) line we're assuming that you're under pool limit, but line above we released the lock one line above. Now at [this](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L142) line we're adding the newly created socket without checkin to see are still under pool limit or not.

In short, we're not doing double checking. Another way of doing this is more intrusive and that is to make use of socket states like for example just before releasing the [lock](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L130) you'd add placeholder socket with `state == connecting`. This way your pool limit would be up to date the moment you release the lock and you'd not need to double check (as you're not adding new socket to the list [here](https://github.com/lyft/mgo/blob/5d0d4e79604660161de357e13cc649cb70c679be/server.go#L142))

rodaine replied:
My concern here is you are introducing a computed variable that must be kept updated with the true `len(server.liveSockets)`. I'd prefer and it's safer going forward to solely rely on `len` instead of a separate field on server. This call is O(1) as the length information is stored on the slice value, so it's not like it's having to walk an array to determine its size or anything.

adilhafeez replied:
Simply relying on len is not going to be enough. There is a right way but that is intrusive and would require more changes in mgo driver. I'm thinking of picking this up after halloween but for now this should work fine.

rodaine commented on server.go:
```diff
@@ -111,36 +112,44 @@ func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (
-		n := len(server.unusedSockets)
-		if poolLimit > 0 && len(server.liveSockets)-n >= poolLimit {
+
+		if poolLimit > 0 && server.activeSocketsCount > poolLimit {
```
`>=`, no?

rodaine commented on server.go:
```diff
@@ -111,36 +112,44 @@ func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (
+			} else {
+				server.Lock()
+				server.activeSocketsCount--
+				server.Unlock()
```
The above 3 diffs (lines 132-152) could be simplified to a single `server.activeSocketCount++` on line 147.

adilhafeez replied:
activeSocketCount is used to protect liveSockets list

rodaine replied:
I get the mechanics, my point is that you can simplify all this code by only mutating this variable once. However, not managing this yourself (as described [here](https://github.com/lyft/mgo/pull/6#discussion_r146630373)) would be better.

rodaine commented on server.go:
```diff
@@ -243,6 +252,7 @@ func (server *mongoServer) AbendSocket(socket *mongoSocket) {
 		return
 	}
 	server.liveSockets = removeSocket(server.liveSockets, socket)
+	server.activeSocketsCount = len(server.liveSockets)
```
Are there any other places where a socket would be released from liveSockets where this value would need to be updated?

adilhafeez replied:
none, have updated all places.

rodaine reviewed (commented):
Awesome work! This would be great to upstream and I think is the root of an issue that plagued me at another company 😅

rodaine commented:
Also, are there tests that we can write to capture this behavioral change?

jsnoop36 commented:
I was also thinking we should try to upstream this.  And was also thinking that it might be worth trying to write a test on this.  But awesome work!!!

adilhafeez commented:
Yep - tests are in progress

adilhafeez committed:
Fix concurrency bug

adilhafeez committed:
add test case and some comments

adilhafeez commented:
I've added a new test to validate this new behavior, if we were to run this test without my fix we'd see following error,

```
➜  mgo git:(lyft-v2-4) ✗ go test -check.v -check.f TestPoolLimitRaceCondition | grep -A 5 c.Assert
OOPS: 0 passed, 1 FAILED
    c.Assert(socketsAlive <= poolLimit, Equals, true,
        Commentf("Pool limit exceeded: limit: %d, socketsAlive: %d", poolLimit, socketsAlive))
... obtained bool = false
... expected bool = true
... Pool limit exceeded: limit: 2, socketsAlive: 3

--
    c.Assert(socketsAlive <= poolLimit, Equals, true,
        Commentf("Pool limit exceeded: limit: %d, socketsAlive: %d", poolLimit, socketsAlive))
... obtained bool = false
... expected bool = true
... Pool limit exceeded: limit: 2, socketsAlive: 3
```

adilhafeez commented:
PTAL @rodaine @jsnoop36

adilhafeez committed:
use socket state to control pool limit

In this code change we're using socket state and a placeholder to keep our slot in aciveSockets slice. This will ensure that we dont run over pool limit.

adilhafeez committed:
fix spacing

rodaine commented on socket.go:
```diff
@@ -38,6 +38,12 @@ import (
 
 type replyFunc func(err error, reply *replyOp, docNum int, docData []byte)
 
+type ScoketState int
```
`SocketState` spelling

rodaine replied:
You can also use a uint8 to (marginally) minimize the space needed for this type

rodaine commented on socket.go:
```diff
@@ -38,6 +38,12 @@ import (
+type ScoketState int
+const (
+	Connecting = 1
+	Connected = 2
```
[Idiomatic enum pattern](https://splice.com/blog/iota-elegant-constants-golang/):

```go
const (
  Connecting SocketState = iota
  Connected
)
```

adilhafeez committed:
fix panic while running tests on mac

[LOG] 14.23654 Socket 0xc4200d25a0 to localhost:40001: killed again: read tcp 127.0.0.1:52570->127.0.0.1:40001: use of closed network connection (previously: Closed explicitly)
... Panic: p3699
f7
f10
 (PC=0x10576F8)

/usr/local/go/src/runtime/panic.go:489
  in gopanic
suite_test.go:157
  in S.pid
suite_test.go:167
  in S.Freeze
cluster_test.go:1086
  in S.TestSocketTimeoutOnInactiveSocket
/usr/local/go/src/reflect/value.go:302
  in Value.Call
/usr/local/go/src/runtime/asm_amd64.s:2197
  in goexit
OOPS: 0 passed, 1 PANICKED
--- FAIL: TestAll (0.11s)
FAIL
exit status 1

adilhafeez committed:
fix spelling and add enum type

adilhafeez commented:
also fixes panic on mac
```
[LOG] 47.81432 Socket 0xc4200d2690 to localhost:40001: killed again: read tcp 127.0.0.1:52305->127.0.0.1:40001: use of closed network connection (previously: Closed explicitly)
... Panic: cannot convert pid to int: 3699
f7
f10 (PC=0x10576F8)
```

lsof command on mac returns more than one line,

```
➜  mgo git:(active_socket_fix) ✗ lsof -iTCP:40001 -sTCP:LISTEN -Fp
p3699
f7
f10
```

adilhafeez committed:
cr
